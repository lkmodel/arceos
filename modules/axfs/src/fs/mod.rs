cfg_if::cfg_if! {
    if #[cfg(feature = "myfs")] {
        pub mod myfs;
        pub const BLOCK_SIZE: usize = 512;
    } else if #[cfg(feature = "fatfs")] {
        pub mod fatfs;
        pub use fatfs::BLOCK_SIZE;
    } else if #[cfg(feature = "ext4fs")] {
        pub mod ext4fs;
        pub use ext4fs::BLOCK_SIZE;
    } else if #[cfg(feature = "ext4_rs")] {

        pub mod ext4;
        pub use ext4::BLOCK_SIZE;
    }
}

// #[cfg(feature = "ext4_rs")]
// pub mod ext4;

#[cfg(feature = "devfs")]
pub use axfs_devfs as devfs;

#[cfg(feature = "ramfs")]
pub use axfs_ramfs as ramfs;
