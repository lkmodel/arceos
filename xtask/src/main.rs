use clap::Parser;
use serde::Deserialize;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::str::from_utf8;
use std::{env, fs, io};

const CC: &str = "riscv64-linux-musl-gcc";
const DYNAMIC_FLAG: [&str; 1] = ["-fPIE", ];
const STATIC_FLAG: [&str; 0] = [];

#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
struct Args {
    /// Architecture to use (e.g., riscv64)
    #[arg(long, default_value = "riscv64")]
    arch: String,

    /// Log level (e.g., warn)
    #[arg(long, short, default_value = "warn")]
    log: String,

    /// Enable QEMU logging (y/n)
    #[arg(long, default_value = "n")]
    qemu_log: String,

    /// App type to run (all, static, dynamic)
    #[arg(long, short)]
    ttype: Option<String>,

    /// Review snapshot and don`t run app
    #[arg(long, short)]
    snapshot: bool,

    /// Which app to run
    app: String,
}

#[derive(Debug, Deserialize)]
struct Config {
    dev: DevConfig,
}

#[derive(Debug, Deserialize)]
struct DevConfig {
    rename: Option<String>,
    ttype: Option<String>,
    dynamic_flags: Option<Vec<String>>,
    static_flags: Option<Vec<String>>,
    snapshot: Option<bool>,
}

fn main() {
    let args = Args::parse();

    if args.snapshot {
        Command::new("cargo")
            .args(&["install", "cargo-insta"])
            .status()
            .expect("Can`t install cargo-insta");
        let current_dir = env::current_dir().expect("Failed to get current directory");
        Command::new("cargo")
            .args(&[
                "insta",
                "review",
                "--workspace-root",
                current_dir.join("payload").join(args.app).to_str().unwrap(),
            ])
            .status()
            .expect("Review failed");
        return;
    }

    let current_dir = env::current_dir().expect("Failed to get current directory");

    // Check and install musl-riscv64
    if !check_installation() {
        if !install_musl_riscv64() {
            eprintln!("Failed to install musl-riscv64");
            return;
        }
    }

    // Checkout the mocklibc branch
    if !check_branch("mocklibc") {
        eprintln!("Failed to switch to mocklibc branch");
        return;
    }

    // Add Rust target
    let status = Command::new("rustup")
        .args(["target", "add", "riscv64gc-unknown-linux-musl"])
        .status()
        .expect("Failed to add Rust target");

    if !status.success() {
        eprintln!("Failed to add Rust target");
        return;
    }

    let is_ci = env::var("CI").is_err();
    if args.app == "all".to_string() || !is_ci {
        traverse_all_app(&current_dir.join("payload"), &|dir: &PathBuf| -> Result<(), String> {
            let arg = Args {
                arch: "riscv64".to_string(),
                log: "warn".to_string(),
                qemu_log: "n".to_string(),
                ttype: None,
                snapshot: false,
                app: dir.iter().last().unwrap().to_str().unwrap().to_string(),
            };
            let config = parse_toml(&current_dir.join("payload").join(&arg.app));
            println!("APP:{}", arg.app);
            dynamic_test(&arg, &current_dir, &config)
        }).expect("Failed to traverse app")
            .iter()
            .for_each(|app| println!("{}\t{}", app[0], app[1]));
        return;
    }

    // Parse the toml file
    let config = parse_toml(&current_dir.join("payload").join(&args.app));
    println!("Config: {:?}", config);
    let mut ttype = config
        .as_ref()
        .and_then(|c| c.dev.ttype.as_ref())
        .unwrap()
        .as_str();
    let args_ttype: &String;
    if args.ttype.is_some() {
        args_ttype = args.ttype.as_ref().unwrap();
        if args_ttype == ttype || ttype == "all" {
            ttype = args_ttype;
        } else {
            eprintln!("Unsupported ttype {}", args_ttype);
            return;
        }
    }

    println!("Architecture: {}", args.arch);
    println!("Log level: {}", args.log);
    println!("QEMU log: {}", args.qemu_log);
    println!("Link type: {}", ttype);
    println!("App: {}", args.app);

    // Build mocklibc
    // let status = Command::new("cargo")
    //     .args([
    //         "build",
    //         "--target",
    //         "riscv64gc-unknown-linux-musl",
    //         "--release",
    //         "-p",
    //         "mocklibc",
    //     ])
    //     .status()
    //     .expect("Failed to build mocklibc");
    //
    // if !status.success() {
    //     eprintln!("Failed to build mocklibc");
    //     return;
    // }

    // Move the built library to the payload directory
    // let lib_path = PathBuf::from("./target/riscv64gc-unknown-linux-musl/release/libmocklibc.so");
    // let payload_path = PathBuf::from(format!("./payload/{}/libmocklibc.so",args.app));
    // std::fs::rename(&lib_path, &payload_path).expect("Failed to move libmocklibc.so");


    // Run tests based on the test type
    match ttype {
        "dynamic" => dynamic_test(&args, &current_dir, &config).unwrap(),
        "static" => static_test(&args, &current_dir, &config).unwrap(),
        "all" => {
            dynamic_test(&args, &current_dir, &config).unwrap();
            println!("----------------------------------------");
            println!(" ");
            println!("----------------------------------------");
            static_test(&args, &current_dir, &config).unwrap();
        }
        _ => eprintln!("Invalid test type"),
    }
}
fn traverse_all_app(path: &PathBuf, cb: &dyn Fn(&PathBuf) -> Result<(), String>) -> io::Result<Vec<Vec<String>>> {
    let mut apps = Vec::new();
    if path.is_dir() {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let path = entry.path();
            if path.is_dir() {
                if let Err(e) = cb(&path) {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), e]);
                } else {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), String::from("OK")]);
                }
            }
        }
    }
    Ok(apps)
}

fn parse_toml(app_path: &PathBuf) -> Option<Config> {
    let toml_path = app_path.join("config.toml");
    if toml_path.exists() {
        let toml_content = fs::read_to_string(toml_path).expect("Failed to read toml file");
        let config: Config = toml::from_str(&toml_content).expect("Failed to parse toml file");
        Some(config)
    } else {
        None
    }
}

fn check_installation() -> bool {
    Command::new("which")
        .arg("riscv64-linux-musl-gcc")
        .output()
        .map(|output| output.status.success())
        .unwrap_or(false)
        || PathBuf::from("/opt/musl_riscv64").exists()
}

fn install_musl_riscv64() -> bool {
    if check_installation() {
        return true;
    }

    if !get_sudo() {
        return false;
    }

    let status = Command::new("git")
        .args(["clone", "https://github.com/richfelker/musl-cross-make.git"])
        .status()
        .expect("Failed to clone musl-cross-make");

    if !status.success() {
        eprintln!("Failed to clone musl-cross-make");
        return false;
    }

    let mut config_mak = std::fs::read_to_string("musl-cross-make/config.mak.dist")
        .expect("Failed to read config.mak.dist");

    config_mak.insert_str(15, "TARGET = riscv64-linux-musl\n");
    config_mak.insert_str(22, "OUTPUT = /opt/musl_riscv64\n");

    std::fs::write("musl-cross-make/config.mak", config_mak).expect("Failed to write config.mak");

    let status = Command::new("make")
        .args(["-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to build musl-cross-make");

    if !status.success() {
        eprintln!("Build failed");
        return false;
    }

    let status = Command::new("sudo")
        .args(["make", "install", "-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to install musl-cross-make");

    if !status.success() {
        eprintln!("Installation failed");
        return false;
    }

    // Add to PATH in .bashrc and .zshrc
    let home_dir = env::var("HOME").expect("Failed to get HOME directory");
    let bashrc_path = PathBuf::from(&home_dir).join(".bashrc");
    let zshrc_path = PathBuf::from(&home_dir).join(".zshrc");

    for rc_file in &[bashrc_path, zshrc_path] {
        if rc_file.exists() {
            let mut rc_content = std::fs::read_to_string(rc_file).expect("Failed to read rc file");
            if !rc_content.contains("/opt/musl_riscv64/bin") {
                rc_content.push_str("\nexport PATH=$PATH:/opt/musl_riscv64/bin\n");
                std::fs::write(rc_file, rc_content).expect("Failed to write rc file");
            }
        }
    }

    println!("Musl RISC-V64 toolchain installation complete");
    std::fs::remove_dir_all("musl-cross-make").expect("Failed to remove musl-cross-make directory");

    // Add musl-riscv64 to PATH
    let mut path = env::var("PATH").unwrap_or_default();
    path.push_str(":/opt/musl_riscv64/bin");
    unsafe {
        env::set_var("PATH", &path);
    }

    true
}

fn get_sudo() -> bool {
    for attempt in 1..=3 {
        println!("Requesting sudo permissions (attempt {}/3)", attempt);
        let status = Command::new("sudo")
            .arg("-v")
            .status()
            .expect("Failed to request sudo permissions");

        if status.success() {
            println!("Sudo permissions granted");
            return true;
        } else {
            println!("Permission denied, please retry");
        }
    }

    eprintln!("Failed to obtain sudo permissions after maximum attempts");
    false
}

fn check_branch(branch_name: &str) -> bool {
    let output = Command::new("git")
        .args(["rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .expect("Failed to get current branch");

    if output.status.success() {
        let current_branch = String::from_utf8(output.stdout).expect("Failed to parse branch name");
        if current_branch.trim() == branch_name {
            println!("Currently on branch {}", branch_name);
            true
        } else {
            println!("Not on branch {}", branch_name);
            let status = Command::new("git")
                .args(["switch", branch_name])
                .status()
                .expect("Failed to switch branch");

            if status.success() {
                println!("Switched to branch {}", branch_name);
                true
            } else {
                eprintln!("Failed to switch branch");
                false
            }
        }
    } else {
        eprintln!("Failed to get current branch");
        false
    }
}

fn static_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, false, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn dynamic_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, true, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn run_make(args: &Args, current_dir: &PathBuf) -> Result<(), String> {
    let status = Command::new("make")
        .args(["defconfig", "ARCH=riscv64"])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make defconfig");

    if !status.success() {
        return Err(String::from("make defconfig failed"));
    }

    let status = Command::new("make")
        .args([
            "A=examples/loader",
            &format!("ARCH={}", args.arch),
            &format!("LOG={}", args.log),
            &format!("QEMU_LOG={}", args.qemu_log),
            "run",
        ])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make");

    if !status.success() {
        return Err(String::from("make run failed"));
    }
    Ok(())
}

/// Generate a binary file from the ELF file
fn generate_bin(elf_file: &str, working_dir: &PathBuf) -> io::Result<()> {
    let elf_path = working_dir.join(elf_file);
    let bin_path = working_dir.join("apps.bin");

    // Create a 32MB empty file
    let file = File::create(&bin_path)?;
    file.set_len(32 * 1024 * 1024)?; // 32MB

    // Get the size of the ELF file
    let app_size = fs::metadata(&elf_path)?.len();

    // Convert the size to a 16-byte hex string and reverse the byte order
    let size_hex = format!("{:016x}", app_size);
    let mut size_bytes = hex::decode(size_hex).unwrap();
    size_bytes.reverse();

    // Write the size to the binary file (first 8 bytes)
    let mut file = OpenOptions::new().write(true).open(&bin_path)?;
    file.write_all(&size_bytes)?;

    // Write the ELF file content to the binary file (starting at offset 8)
    let mut elf_file = File::open(elf_path)?;
    let mut buffer = Vec::new();
    elf_file.read_to_end(&mut buffer)?;

    file.seek(SeekFrom::Start(8))?;
    file.write_all(&buffer)?;

    // Copy the binary file to the parent directory
    let parent_bin_path = working_dir.parent().unwrap().join("apps.bin");
    fs::rename(&bin_path, &parent_bin_path)?;
    Ok(())
}

/// Build a ELF file
fn build(elf_path: &PathBuf, ttype: bool, config: &Option<Config>) -> io::Result<()> {
    let elf_file = elf_path
        .iter()
        .last()
        .unwrap()
        .to_str()
        .unwrap()
        .split('_')
        .next()
        .unwrap()
        .to_string();
    // Determine the rename value from the config
    let rename = config
        .as_ref()
        .and_then(|c| c.dev.rename.as_ref())
        .unwrap_or(&elf_file)
        .as_str();
    let c_file = format!("{}.c", rename); // C source file
    let elf_output = format!("{}", rename); // Output ELF file

    // Determine the flags based on the config
    let dynamic_flags = config
        .as_ref()
        .and_then(|c| c.dev.dynamic_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| DYNAMIC_FLAG.iter().map(|s| *s).collect());

    let static_flags = config
        .as_ref()
        .and_then(|c| c.dev.static_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| STATIC_FLAG.iter().map(|s| *s).collect());

    // Compile the C file
    let mut status = Command::new(CC);
    let flags = if ttype {
        dynamic_flags
    } else {
        static_flags
    };
    status
        .args(&["-v", &c_file])
        .args(flags)
        .args(&["-o", &elf_output])
        .current_dir(elf_path)
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    let re = status.status()?;

    if !re.success() {
        return Err(io::Error::new(
            io::ErrorKind::Other,
            "Failed to compile C file",
        ));
    }

    let mut insta_set = insta::Settings::clone_current();
    insta_set.set_snapshot_path(elf_path.join("snapshot"));
    insta_set.set_prepend_module_to_snapshot(false);

    let t_type: String;
    if ttype {
        t_type = "dynamic".to_string();
    } else {
        t_type = "static".to_string();
    }

    // Generate disassembly file
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();

    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.S", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(output_file.join(&format!("{}.S", elf_file)), output.stdout)
        .expect("Failed to write ELF file");

    // Generate ELF info file
    let output = Command::new("riscv64-linux-musl-readelf")
        .args(&["-a", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-readelf");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.elf", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.elf", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate full disassembly and symbol table
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-x", "-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.dump", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.dump", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate the binary file
    generate_bin(&elf_output, &elf_path)?;

    Ok(())
}

