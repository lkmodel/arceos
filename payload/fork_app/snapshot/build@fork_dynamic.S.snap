---
source: xtask/src/main.rs
description: riscv64-linux-musl-gcc (GCC) 11.2.1 20211120
expression: from_utf8(&output.stdout).unwrap()
---
fork:     file format elf64-littleriscv


Disassembly of section .plt:

0000000000000590 <.plt>:
 590:	00002397          	auipc	t2,0x2
 594:	41c30333          	sub	t1,t1,t3
 598:	a783be03          	ld	t3,-1416(t2) # 2008 <__TMC_END__>
 59c:	fd430313          	addi	t1,t1,-44
 5a0:	a7838293          	addi	t0,t2,-1416
 5a4:	00135313          	srli	t1,t1,0x1
 5a8:	0082b283          	ld	t0,8(t0)
 5ac:	000e0067          	jr	t3

00000000000005b0 <printf@plt>:
 5b0:	00002e17          	auipc	t3,0x2
 5b4:	a68e3e03          	ld	t3,-1432(t3) # 2018 <printf>
 5b8:	000e0367          	jalr	t1,t3
 5bc:	00000013          	nop

00000000000005c0 <fork@plt>:
 5c0:	00002e17          	auipc	t3,0x2
 5c4:	a60e3e03          	ld	t3,-1440(t3) # 2020 <fork>
 5c8:	000e0367          	jalr	t1,t3
 5cc:	00000013          	nop

00000000000005d0 <__libc_start_main@plt>:
 5d0:	00002e17          	auipc	t3,0x2
 5d4:	a58e3e03          	ld	t3,-1448(t3) # 2028 <__libc_start_main>
 5d8:	000e0367          	jalr	t1,t3
 5dc:	00000013          	nop

Disassembly of section .text:

00000000000005e0 <_start>:
 5e0:	00002197          	auipc	gp,0x2
 5e4:	22018193          	addi	gp,gp,544 # 2800 <__global_pointer$>
 5e8:	850a                	mv	a0,sp
 5ea:	00002597          	auipc	a1,0x2
 5ee:	85658593          	addi	a1,a1,-1962 # 1e40 <_DYNAMIC>
 5f2:	ff017113          	andi	sp,sp,-16
 5f6:	a009                	j	5f8 <_start_c>

00000000000005f8 <_start_c>:
 5f8:	410c                	lw	a1,0(a0)
 5fa:	00850613          	addi	a2,a0,8
 5fe:	4781                	li	a5,0
 600:	00002717          	auipc	a4,0x2
 604:	a6873703          	ld	a4,-1432(a4) # 2068 <_fini>
 608:	00002697          	auipc	a3,0x2
 60c:	a386b683          	ld	a3,-1480(a3) # 2040 <_init>
 610:	00002517          	auipc	a0,0x2
 614:	a5053503          	ld	a0,-1456(a0) # 2060 <_GLOBAL_OFFSET_TABLE_+0x30>
 618:	bf65                	j	5d0 <__libc_start_main@plt>

000000000000061a <deregister_tm_clones>:
 61a:	00002517          	auipc	a0,0x2
 61e:	9ee50513          	addi	a0,a0,-1554 # 2008 <__TMC_END__>
 622:	00002797          	auipc	a5,0x2
 626:	9e678793          	addi	a5,a5,-1562 # 2008 <__TMC_END__>
 62a:	00a78863          	beq	a5,a0,63a <deregister_tm_clones+0x20>
 62e:	00002797          	auipc	a5,0x2
 632:	a2a7b783          	ld	a5,-1494(a5) # 2058 <_ITM_deregisterTMCloneTable>
 636:	c391                	beqz	a5,63a <deregister_tm_clones+0x20>
 638:	8782                	jr	a5
 63a:	8082                	ret

000000000000063c <register_tm_clones>:
 63c:	00002517          	auipc	a0,0x2
 640:	9cc50513          	addi	a0,a0,-1588 # 2008 <__TMC_END__>
 644:	00002597          	auipc	a1,0x2
 648:	9c458593          	addi	a1,a1,-1596 # 2008 <__TMC_END__>
 64c:	8d89                	sub	a1,a1,a0
 64e:	4035d793          	srai	a5,a1,0x3
 652:	91fd                	srli	a1,a1,0x3f
 654:	95be                	add	a1,a1,a5
 656:	8585                	srai	a1,a1,0x1
 658:	c599                	beqz	a1,666 <register_tm_clones+0x2a>
 65a:	00002797          	auipc	a5,0x2
 65e:	9f67b783          	ld	a5,-1546(a5) # 2050 <_ITM_registerTMCloneTable>
 662:	c391                	beqz	a5,666 <register_tm_clones+0x2a>
 664:	8782                	jr	a5
 666:	8082                	ret

0000000000000668 <__do_global_dtors_aux>:
 668:	1141                	addi	sp,sp,-16
 66a:	e022                	sd	s0,0(sp)
 66c:	00002417          	auipc	s0,0x2
 670:	a0c40413          	addi	s0,s0,-1524 # 2078 <completed.1>
 674:	00044783          	lbu	a5,0(s0)
 678:	e406                	sd	ra,8(sp)
 67a:	eb95                	bnez	a5,6ae <__do_global_dtors_aux+0x46>
 67c:	00002797          	auipc	a5,0x2
 680:	9bc7b783          	ld	a5,-1604(a5) # 2038 <__cxa_finalize>
 684:	c791                	beqz	a5,690 <__do_global_dtors_aux+0x28>
 686:	00002517          	auipc	a0,0x2
 68a:	97a53503          	ld	a0,-1670(a0) # 2000 <__dso_handle>
 68e:	9782                	jalr	a5
 690:	f8bff0ef          	jal	ra,61a <deregister_tm_clones>
 694:	00002797          	auipc	a5,0x2
 698:	9b47b783          	ld	a5,-1612(a5) # 2048 <__deregister_frame_info>
 69c:	c791                	beqz	a5,6a8 <__do_global_dtors_aux+0x40>
 69e:	00000517          	auipc	a0,0x0
 6a2:	22250513          	addi	a0,a0,546 # 8c0 <__FRAME_END__>
 6a6:	9782                	jalr	a5
 6a8:	4785                	li	a5,1
 6aa:	00f40023          	sb	a5,0(s0)
 6ae:	60a2                	ld	ra,8(sp)
 6b0:	6402                	ld	s0,0(sp)
 6b2:	0141                	addi	sp,sp,16
 6b4:	8082                	ret

00000000000006b6 <frame_dummy>:
 6b6:	00002797          	auipc	a5,0x2
 6ba:	9ba7b783          	ld	a5,-1606(a5) # 2070 <__register_frame_info>
 6be:	cf99                	beqz	a5,6dc <frame_dummy+0x26>
 6c0:	1141                	addi	sp,sp,-16
 6c2:	e406                	sd	ra,8(sp)
 6c4:	00002597          	auipc	a1,0x2
 6c8:	9bc58593          	addi	a1,a1,-1604 # 2080 <object.0>
 6cc:	00000517          	auipc	a0,0x0
 6d0:	1f450513          	addi	a0,a0,500 # 8c0 <__FRAME_END__>
 6d4:	9782                	jalr	a5
 6d6:	60a2                	ld	ra,8(sp)
 6d8:	0141                	addi	sp,sp,16
 6da:	b78d                	j	63c <register_tm_clones>
 6dc:	b785                	j	63c <register_tm_clones>

00000000000006de <main>:
 6de:	7179                	addi	sp,sp,-48
 6e0:	f406                	sd	ra,40(sp)
 6e2:	f022                	sd	s0,32(sp)
 6e4:	1800                	addi	s0,sp,48
 6e6:	878a                	mv	a5,sp
 6e8:	fef43423          	sd	a5,-24(s0)
 6ec:	87a2                	mv	a5,s0
 6ee:	fef43023          	sd	a5,-32(s0)
 6f2:	8786                	mv	a5,ra
 6f4:	fcf43c23          	sd	a5,-40(s0)
 6f8:	fd843683          	ld	a3,-40(s0)
 6fc:	fe043603          	ld	a2,-32(s0)
 700:	fe843583          	ld	a1,-24(s0)
 704:	00000517          	auipc	a0,0x0
 708:	0d450513          	addi	a0,a0,212 # 7d8 <main+0xfa>
 70c:	ea5ff0ef          	jal	ra,5b0 <printf@plt>
 710:	00000597          	auipc	a1,0x0
 714:	0f858593          	addi	a1,a1,248 # 808 <main+0x12a>
 718:	00000517          	auipc	a0,0x0
 71c:	10050513          	addi	a0,a0,256 # 818 <main+0x13a>
 720:	e91ff0ef          	jal	ra,5b0 <printf@plt>
 724:	e9dff0ef          	jal	ra,5c0 <fork@plt>
 728:	87aa                	mv	a5,a0
 72a:	fcf42a23          	sw	a5,-44(s0)
 72e:	fd442783          	lw	a5,-44(s0)
 732:	85be                	mv	a1,a5
 734:	00000517          	auipc	a0,0x0
 738:	0f450513          	addi	a0,a0,244 # 828 <main+0x14a>
 73c:	e75ff0ef          	jal	ra,5b0 <printf@plt>
 740:	fd442783          	lw	a5,-44(s0)
 744:	2781                	sext.w	a5,a5
 746:	0007dc63          	bgez	a5,75e <main+0x80>
 74a:	fd442783          	lw	a5,-44(s0)
 74e:	85be                	mv	a1,a5
 750:	00000517          	auipc	a0,0x0
 754:	0e850513          	addi	a0,a0,232 # 838 <main+0x15a>
 758:	e59ff0ef          	jal	ra,5b0 <printf@plt>
 75c:	a805                	j	78c <main+0xae>
 75e:	fd442783          	lw	a5,-44(s0)
 762:	2781                	sext.w	a5,a5
 764:	eb99                	bnez	a5,77a <main+0x9c>
 766:	fd442783          	lw	a5,-44(s0)
 76a:	85be                	mv	a1,a5
 76c:	00000517          	auipc	a0,0x0
 770:	0e450513          	addi	a0,a0,228 # 850 <main+0x172>
 774:	e3dff0ef          	jal	ra,5b0 <printf@plt>
 778:	a811                	j	78c <main+0xae>
 77a:	fd442783          	lw	a5,-44(s0)
 77e:	85be                	mv	a1,a5
 780:	00000517          	auipc	a0,0x0
 784:	10050513          	addi	a0,a0,256 # 880 <main+0x1a2>
 788:	e29ff0ef          	jal	ra,5b0 <printf@plt>
 78c:	00000597          	auipc	a1,0x0
 790:	07c58593          	addi	a1,a1,124 # 808 <main+0x12a>
 794:	00000517          	auipc	a0,0x0
 798:	11c50513          	addi	a0,a0,284 # 8b0 <main+0x1d2>
 79c:	e15ff0ef          	jal	ra,5b0 <printf@plt>
 7a0:	878a                	mv	a5,sp
 7a2:	fef43423          	sd	a5,-24(s0)
 7a6:	87a2                	mv	a5,s0
 7a8:	fef43023          	sd	a5,-32(s0)
 7ac:	8786                	mv	a5,ra
 7ae:	fcf43c23          	sd	a5,-40(s0)
 7b2:	fd843683          	ld	a3,-40(s0)
 7b6:	fe043603          	ld	a2,-32(s0)
 7ba:	fe843583          	ld	a1,-24(s0)
 7be:	00000517          	auipc	a0,0x0
 7c2:	01a50513          	addi	a0,a0,26 # 7d8 <main+0xfa>
 7c6:	debff0ef          	jal	ra,5b0 <printf@plt>
 7ca:	4781                	li	a5,0
 7cc:	853e                	mv	a0,a5
 7ce:	70a2                	ld	ra,40(sp)
 7d0:	7402                	ld	s0,32(sp)
 7d2:	6145                	addi	sp,sp,48
 7d4:	8082                	ret
